src/openrct2/ride/Ride.cpp

```diff
@@ -3026,19 +3026,98 @@ static void RideOpenBlockBrakes(const CoordsXYE& startElement)
         auto trackType = currentElement.element->AsTrack()->GetTrackType();
         switch (trackType)
         {
+            case TrackElemType::BlockBrakes:
+                BlockBrakeSetLinkedBrakesClosed(
+                    CoordsXYZ(currentElement.x, currentElement.y, currentElement.element->GetBaseZ()),
+                    *currentElement.element->AsTrack(), false);
+                [[fallthrough]];
             case TrackElemType::EndStation:
             case TrackElemType::CableLiftHill:
             case TrackElemType::Up25ToFlat:
             case TrackElemType::Up60ToFlat:
             case TrackElemType::DiagUp25ToFlat:
             case TrackElemType::DiagUp60ToFlat:
-            case TrackElemType::BlockBrakes:
                 currentElement.element->AsTrack()->SetBrakeClosed(false);
                 break;
         }
     } while (track_block_get_next(&currentElement, &currentElement, nullptr, nullptr)
              && currentElement.element != startElement.element);
 }
+void BrakeLinkToBlockBrake(const CoordsXYZ& vehicleTrackLocation, TrackElement& brake)
+{
+    CoordsXYE output = CoordsXYE(vehicleTrackLocation.x, vehicleTrackLocation.y, reinterpret_cast<TileElement*>(&brake));
+    int32_t outputZ = vehicleTrackLocation.z;
+    do
+    {
+        if (output.element->AsTrack()->GetTrackType() == TrackElemType::BlockBrakes)
+        {
+            brake.SetBrakeClosed(
+                (brake.GetBrakeBoosterSpeed() >= output.element->AsTrack()->GetBrakeBoosterSpeed())
+                || (output.element->AsTrack()->IsBrakeClosed()));
```

Comment:
This can be much clearer. 

```
const auto blockBrakeSpeedIsLower = output.element->AsTrack()->GetBrakeBoosterSpeed() < brake.GetBrakeBoosterSpeed();
const auto blockBrakeIsClosed = output.element->AsTrack()->IsBrakeClosed();
const auto closeBrake = blockBrakeSpeedIsLower || blockBrakeIsClosed;